### ISSUE TITLE
Support for other transports (SPI, BLE, TCP/IP, ...)

### ISSUE THREAD

**Issue by pperanich** [jamesmunns/postcard-rpc#26]
Thank you for the fantastic work on this library! It significantly reduces the effort to get RPC comms established between a host and embedded device.

Currently, the framework as a whole only supports transport over serial or USB, with the `target_server` side specifically relying on the embassy-usb driver. I am interested in what it would take to support other transports such as SPI, BLE (using the `trouble` lib, which abstracts over Bluetooth HCI and is thus more versatile than implementing support for a specific stack like the NRF SoftDevice), or TCP/IP via `smoltcp`. While I see a clear path to implementing SPI or BLE support on the `host_client` side, it is unclear how to extend this support to the `target_server` embedded side. I've loosely been following your blog where it looks like [you've been thinking about this](https://onevariable.com/blog/thoughts-netstack/) on some level already. I hope this issue can serve as a log or discussion place for these implementations.

**Comment by jamesmunns** [comment]
Yep, on the PC side, https://github.com/jamesmunns/postcard-rpc/pull/25 is starting to work towards this, and in the near future I plan to restore "UART + COBS" functionality.

Likely, we will need some kind of similar "Wire" abstraction for the MCU side as well, and I'll need to rework the `define_dispatch!` macro to be less hardcoded to USB, and instead require you to name some type where `T: Wire` or something like that.

I'm open to PRs that make this happen, ideally starting with adding back UART+COBS, but also open to a PR that refactors this in a way where supporting two different transports (ideally without a lot of duplicated dispatcher code) is possible.

**Comment by okhsunrog** [comment]
It would be awesome if we could use any transport that implements embedded-io

**Comment by jamesmunns** [comment]
If anyone is following this, the changes between 0.7 and 0.10 have made it much easier to write a transport, including blanket ones like @okhsunrog suggested.

I probably will not write this in the near future (as I don't need it), but if someone wants to take it on, I'm happy to chat or say how I would do it.

**Comment by jamesmunns** [comment]
Added RTT to the title because at some point I would like to have that.

**Comment by MathiasKoch** [comment]
+1 for RTT transport support :+1:

**Comment by okhsunrog** [comment]
I'm using a cli terminal in a similar way, I implemented embedded-io for RTT and I can easily switch interfaces. I'm actually interested in trying postcard-rpc with RTT, but I'm but sure yet if it fits my task.
@jamesmunns can it be used with 2 embedded devices, both being no_std or the host side needs to be std?

**Comment by jamesmunns** [comment]
@okhsunrog:

one potential blocker to supporting RTT is we need some kind of async `recv()` to be notified when a frame is ready. We could have a low priority, blocking task that does this (using interrupt executors on embassy), but it might be nicer to have some way for the debugger to "notify" when it has written RTT data, potentially using an interrupt as a callback handler. I discussed this in chat today: https://libera.irclog.whitequark.org/rust-embedded/2024-11-01#1730467273-1730468273; - there were some ideas of how this could be possible

re: two embedded devices, see #37 - there isn't a no-std `client` yet, only a std/host `client`. It should definitely be possible to implement, especially if you are willing to accept some limitations (you can't clone the client, only one in-flight request at a time, handling topic subscriptions might be a little tricky), but this isn't a priority for me at the moment.

**Comment by okhsunrog** [comment]
@jamesmunns I guess we can do something like this for now, polling in a loop:
```
pub async fn async_recv(channels: &mut Channels, buf: &mut [u8]) -> usize {
    loop {
        let n = channels.down.0.read(buf);
        if n > 0 {
            return n;
        }
        Timer::after(Duration::from_millis(1)).await;
    }
}

```

**Comment by jamesmunns** [comment]
@okhsunrog yep, I agree that is a reasonable "not great, not bad" solution, and is exactly what I had in mind :)

**Comment by Dicklessgreat** [comment]
I'm developing an embedded device that should be operable through multiple interfaces:
- Via USB from a host PC with rich GUI
- Via physical controls (like rotary encoders) directly on the device
- (Potentially via other interfaces in the future)

The goal is to make all device functionality accessible regardless of the available interface. This leads to two concerns with the current implementation:

1. For endpoints and topics, we need to duplicate their definitions in each "Wire", which increases maintenance costs.

2. In my embedded application, I want to post messages to the server from internal tasks (like UI events from physical controls). Currently, this seems difficult because the server is tightly coupled with external transport implementations (like USB). Would it make sense to have an "internal" transport implementation that could work alongside external transports?

Let me know if you'd like more details about either of these use cases.

**Comment by jamesmunns** [comment]
Hey @Dicklessgreat, so this issue is *mostly* for tracking the implementation of new `WireRx`/`WireTx` impls. You asked two main things (let me know if I missed anything):

1. "Can we offer multiple interfaces", and the answer is "not yet, but maybe in the future". I've opened #67 to track this. This feature is something I am open to, and if anyone is interested in sponsoring this work, I'd be happy to prioritize it!
2. "Can we have local/internal clients", and the answer is "I don't think this is a good idea". I've opened #68 for this, and I'm open to discuss this there.

**Comment by jamesmunns** [comment]
Noting here, I currently have two WIP transports, help welcome for getting them polished up and merged:

* `embedded-io`/UART support: https://github.com/jamesmunns/postcard-rpc/pull/83
* RTT support via probe-rs: https://github.com/jamesmunns/prpcrtt
